src-control/Control/RefCount.hs

```diff
@@ -89,6 +92,38 @@ removeReference RefCounter{countVar, finaliser} = mask_ $ do
     assertWithCallStack (prevCount > 0) $ pure ()
     when (prevCount == 1) $ sequence_ finaliser
 
+{-# SPECIALISE removeReferenceN :: HasCallStack => RefCounter IO -> Word64 -> IO () #-}
+-- | Decrease the reference counter by @n@. @n@ must be a positive number.
+--
+-- The count must be known (from context) to be non-zero and at least as large
+-- as @n@. Typically this will be because the caller has @n@ references already
+-- (that they took out themselves or were given).
+removeReferenceN :: (HasCallStack, PrimMonad m, MonadMask m) => RefCounter m -> Word64 -> m ()
+removeReferenceN RefCounter{countVar, finaliser} n = mask_ $ do
+    -- n should be positive
+    assert (n > 0) $ pure ()
+    let !n' = fromIntegralChecked n
+    prevCount <- fetchSubInt countVar n'
+    -- the reference count must not already be 0, because then the finaliser
+    -- will have run already
+    assertWithCallStack (prevCount > 0) $ pure ()
+    -- the reference count can not go below zero
+    assertWithCallStack (prevCount >= n') $ pure ()
+    when (prevCount <= n') $ sequence_ finaliser
```

Comment:
This runs the finaliser even if the reference count goes below zero. Is this necessary? Probably yes, but we can experiment